File: src/test/java/com/runninglane/dto/buddy/test/java/cases/bytecode/NonPropertyAbstractMethodsHandlingTest.java
package com.runninglane.dto.buddy.test.java.cases.bytecode;

import com.runninglane.dto.buddy.javainterop.DtoBuddy;
import com.runninglane.dto.buddy.javainterop.bytecode.ByteCodeStrategy;
import com.runninglane.dto.buddy.javainterop.bytecode.ThreeStepsByteCodeStrategy;
import com.runninglane.dto.buddy.javainterop.bytecode.compile.CompileJavaByteCodeStrategyCompliment;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeSpec;
import kotlin.reflect.KClass;
import kotlin.reflect.KFunction;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.junit.jupiter.api.Test;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class NonPropertyAbstractMethodsHandlingTest {
    public static abstract class TestBaseClass {
        protected String greeting = "Hello";

        public String getGreeting() {
            return greeting;
        }

        public void setGreeting(String greeting) {
            this.greeting = greeting;
        }

        public abstract String shout(String name);
    }

    static class TestByteCodeStrategyCompliment extends CompileJavaByteCodeStrategyCompliment {
        @Override
        public @NotNull TypeSpec.Builder handleNonPropertyAbstractMethods(
            @NotNull TypeSpec.Builder builder,
            @NotNull List<Method> methods,
            @Nullable Map<String, ? extends Class<?>> typeParamsMapByName
        ) {
            TypeSpec.Builder updatedBuilder = builder.addMethod(
                MethodSpec.methodBuilder("shout")
                    .addAnnotation(Override.class)
                    .addModifiers(javax.lang.model.element.Modifier.PUBLIC)
                    .addParameter(String.class, "name")
                    .returns(String.class)
                    .addCode("return getGreeting() + \" \" + name + \"!\";\n")
                    .build()
            );

            List<Method> unhandledMethods = methods.stream()
                .filter(method -> !method.getName().equals("shout"))
                .collect(Collectors.toList());

            return super.handleNonPropertyAbstractMethods(updatedBuilder, unhandledMethods, typeParamsMapByName);
        }
    }

    @Test
    public void testHandleNonPropertyAbstractMethods() {
        ByteCodeStrategy byteCodeStrategy = new ThreeStepsByteCodeStrategy<>(
            new TestByteCodeStrategyCompliment()
        );
        DtoBuddy dtoBuddy = new DtoBuddy(byteCodeStrategy);
        Class<?> concreteClass = dtoBuddy.implement(TestBaseClass.class, null);
        Method shoutMethod = Arrays.stream(concreteClass.getMethods())
            .filter(method -> method.getName().equals("shout"))
            .findFirst()
            .orElseThrow();
        assert !Modifier.isAbstract(shoutMethod.getModifiers());
        TestBaseClass dto = dtoBuddy.create(concreteClass);
        dto.setGreeting("Hello");
        assert "Hello John!".equals(dto.shout("John"));

        // Test with a different greeting
        dto.setGreeting("Hi");
        assert "Hi John!".equals(dto.shout("John"));
    }
}

File: src/test/kotlin/com/runninglane/dto/buddy/test/cases/bytecode/NonPropertyAbstractMethodsHandlingTest.kt
package com.runninglane.dto.buddy.test.cases.bytecode

import com.runninglane.dto.buddy.DtoBuddy
import com.runninglane.dto.buddy.bytecode.k2jvm.EmbeddedCompilerByteCodeStrategy
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.TypeSpec
import org.junit.jupiter.api.Assertions.assertFalse
import java.lang.reflect.Modifier
import kotlin.reflect.KFunction
import kotlin.reflect.full.functions
import kotlin.test.Test

class NonPropertyAbstractMethodsHandlingTest {
    abstract class TestBaseClass {
        open var greeting: String = "Hello"
        abstract fun shout(name: String): String
    }

    class TestByteCodeStrategy : EmbeddedCompilerByteCodeStrategy() {
        override fun handleNonPropertyAbstractFunctions(
            builder: TypeSpec.Builder,
            functions: List<KFunction<*>>
        ): TypeSpec.Builder {
            val updatedBuilder = builder.addFunction(
                FunSpec.builder("shout")
                    .addModifiers(com.squareup.kotlinpoet.KModifier.OVERRIDE)
                    .addParameter("name", String::class)
                    .returns(String::class)
                    .addStatement("return \"\${greeting} \$name!\"")
                    .build()
            )
            val unhandledMethods = functions.filter { it.name != "shout" }
            return super.handleNonPropertyAbstractFunctions(updatedBuilder, unhandledMethods)
        }
    }

    @Test
    fun testHandleNonPropertyAbstractMethods() {
        val dtoBuddy = DtoBuddy(TestByteCodeStrategy())
        val concreteClass = dtoBuddy.implement(TestBaseClass::class)
        val shoutFunction = concreteClass.functions.first { it.name == "shout" }
        assertFalse(shoutFunction.isAbstract)
        val dto = dtoBuddy.create<TestBaseClass>(concreteClass)
        dto.greeting = "Hello"
        assert(dto.shout("John") == "Hello John!")

        // Test with a different greeting
        dto.greeting = "Hi"
        assert(dto.shout("John") == "Hi John!")
    }
}